### PR TITLE
io, stream: assert !Unpin for ext trait futures

### PR DESCRIPTION
**Pull request by taiki-e**
## Motivation

https://github.com/tokio-rs/tokio/pull/2910#pullrequestreview-501739040

### CODE REVIEW

File: tokio/tests/async_send_sync.rs [hunk 1]
```diff
@@ -252,3 +284,59 @@ async_assert_fn!(tokio::time::timeout_at(Instant, BoxFutureSync<()>): Send & Syn
...
 async_assert_fn!(tokio::time::timeout_at(Instant, BoxFuture<()>): !Send & !Sync);
 async_assert_fn!(tokio::time::Interval::tick(_): Send & Sync);
+
+async_assert_fn!(tokio::stream::StreamExt::next(&mut BoxStream<()>): !Unpin);
+async_assert_fn!(tokio::stream::StreamExt::try_next(&mut BoxStream<Result<(), ()>>): !Unpin);
+async_assert_fn!(tokio::stream::StreamExt::all(&mut BoxStream<()>, fn(())->bool): !Unpin);
+async_assert_fn!(tokio::stream::StreamExt::any(&mut BoxStream<()>, fn(())->bool): !Unpin);
+async_assert_fn!(tokio::stream::StreamExt::fold(&mut BoxStream<()>, (), fn((), ())->()): !Unpin);
+// TODO: error: type annotations needed for `stream::collect::Collect<&mut Pin<Box<(dyn Stream<Item = ()> + 'static)>>, T>`
+// async_assert_fn!(tokio::stream::StreamExt::collect(&mut BoxStream<()>): !Unpin);
```
Comment (taiki-e) [inline tokio/tests/async_send_sync.rs:294]: The full error message is:

```text

error[E0282]: type annotations needed for `stream::collect::Collect<&mut Pin<Box<(dyn Stream<Item = ()> + 'static)>>, T>`
   --> tokio/tests/async_send_sync.rs:139:23
    |
139 |             let f = $($f $(::<$($generic),*>)? )::+( $( into_todo!($arg) ),* );
    |                 -     ^^ cannot infer type for type parameter `T` declared on the associated function `collect`
    |                 |
    |                 consider giving `f` the explicit type `stream::collect::Collect<&mut Pin<Box<(dyn Stream<Item = ()> + 'static)>>, T>`, where the type parameter `T` is specified
...
294 | async_assert_fn!(tokio::stream::StreamExt::collect(&mut BoxStream<()>): !Unpin);
    | -------------------------------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

Comment (Darksonn) [reply tokio/tests/async_send_sync.rs:294]: This compiles
```rust
async_assert_fn!(tokio::stream::StreamExt::collect<Vec<()>>(&mut BoxStream<()>): !Unpin);
```

**Review by Darksonn — APPROVED** [review]
Seems good to me.

File: tokio/tests/async_send_sync.rs (hunk 1 above)
Comment (taiki-e) [reply tokio/tests/async_send_sync.rs]: Thanks

**Pull request merged**